agent/engine/docker_task_engine_linux_test.go

```diff
@@ -1057,3 +1057,219 @@ func TestPauseContainerWitServiceConnect(t *testing.T) {
+				verifyServiceConnectPauseContainerBridgeMode(t, ctx, config, hostConfig, y, z)
+			}).Return(dockerapi.DockerContainerMetadata{DockerID: pauseContainerID}),
+		dockerClient.EXPECT().StartContainer(gomock.Any(), pauseContainerID, defaultConfig.ContainerStartTimeout).Return(
+			dockerapi.DockerContainerMetadata{
+				DockerID: pauseContainerID,
+				NetworkSettings: &types.NetworkSettings{
+					DefaultNetworkSettings: types.DefaultNetworkSettings{IPAddress: "1.2.3.4"},
+				}},
+		),
+	)
+
+	// For SC and sleep container
+	imageManager.EXPECT().AddAllImageStates(gomock.Any()).AnyTimes()
+	dockerClient.EXPECT().PullImage(gomock.Any(), gomock.Any(), nil, gomock.Any()).Return(dockerapi.DockerContainerMetadata{}).Times(2)
+	imageManager.EXPECT().RecordContainerReference(gomock.Any()).Return(nil).Times(2)
+	imageManager.EXPECT().GetImageStateFromImageName(gomock.Any()).Return(nil, false).Times(2)
+	dockerClient.EXPECT().APIVersion().Return(defaultDockerClientAPIVersion, nil).Times(2)
+
+	dockerClient.EXPECT().CreateContainer(gomock.Any(), gomock.Any(), gomock.Any(),
+		gomock.Any(), gomock.Any()).Return(dockerapi.DockerContainerMetadata{DockerID: scContainerID})
+	dockerClient.EXPECT().CreateContainer(gomock.Any(), gomock.Any(), gomock.Any(),
+		gomock.Any(), gomock.Any()).Return(dockerapi.DockerContainerMetadata{DockerID: sleepContainerID})
+
+	// don't need to specify HEALTHY here because it will get another stage transition to update metadata
+	dockerClient.EXPECT().StartContainer(gomock.Any(), scContainerID, defaultConfig.ContainerStartTimeout).Return(
+		dockerapi.DockerContainerMetadata{DockerID: scContainerID})
+	dockerClient.EXPECT().StartContainer(gomock.Any(), sleepContainerID, defaultConfig.ContainerStartTimeout).Return(
+		dockerapi.DockerContainerMetadata{DockerID: sleepContainerID})
+
+	// SC Container gets an additional inspect call when transitioning to RESOURCE_PROVISIONED, and we need to include
+	// health info here to update metadata correctly
+	dockerClient.EXPECT().InspectContainer(gomock.Any(), scContainerID, gomock.Any()).Return(
+		&types.ContainerJSON{
+			ContainerJSONBase: &types.ContainerJSONBase{
+				ID: scContainerID,
+				State: &types.ContainerState{
+					Pid:    containerPid,
+					Health: &types.Health{Status: types.Healthy},
+				},
+			}}, nil)
+
+	cleanup := make(chan time.Time)
+	defer close(cleanup)
+	mockTime.EXPECT().Now().Return(time.Now()).MinTimes(1)
+	dockerClient.EXPECT().DescribeContainer(gomock.Any(), scContainerID).AnyTimes()
+	dockerClient.EXPECT().DescribeContainer(gomock.Any(), sleepContainerID).AnyTimes()
+	dockerClient.EXPECT().DescribeContainer(gomock.Any(), pauseContainerID).AnyTimes()
+
+	err = taskEngine.Init(ctx)
+	assert.NoError(t, err)
+	taskEngine.AddTask(sleepTask)
+	stateChangeEvents := taskEngine.StateChangeEvents()
+	verifyTaskIsRunning(stateChangeEvents, sleepTask)
+
+	mockTime.EXPECT().After(gomock.Any()).Return(cleanup).MinTimes(1)
+	gomock.InOrder(
```

Comment:
can we also verify here that the task container is stopped first?

Reply:
In this commit we simulate a docker container stop event for task container in order to start the task stop workflow, which means there isn't actually going to be a `StopContainer` call for it. I've updated the test in the newer commit to update task desired status to `STOPPED` instead so that we can verify what you mentioned.